### PR TITLE
Design docs v2

### PR DESCRIPTION
**Pull request by samuel-hawker**

### CODE REVIEW

File: 007-design-docs.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
+# Design Documentation Proposal
+
+This document is intended to progress discussion about how to write and maintain [design documentation](https://en.wikipedia.org/wiki/Software_design_description) for Strimzi.
+
+The main goal of design documents for Strimzi is to make contributors more effective by encouraging all involved to think through the design and gather feedback from others. People often think the point of a design doc is to to teach others about some system or serve as documentation later on. While those can be beneficial side effects, they are not the goal in and of themselves.
```
Comment (katheris) [inline 007-design-docs.md:5]: Is this still the main goal? It still seems at odds to the benefits listed

File: 007-design-docs.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+```
+The Strimzi cluster-operator is a component that manages user specified components in the Kubernetes cluster. It achieves this by reading specifications from custom resource objects.... (Continued)
+
+The cluster-operator is split into multiple operators that are managed by [Vertx](link), these operators are:
+- [Kafka Operator](../kafka-operator.md)
+- [KafkaUser Operator](../kafka-user-operator.md)
+(Continued)
+```
+
+The above is just an example but I think it demonstrates the value of a deeper context for how the Strimzi project is written but without general specifics. Moreover this should be fairly simple to maintain as the overall architecture and flow doesn't change often and should need little more than a single sentence change when they do.
```
Comment (katheris) [inline 007-design-docs.md:38]: "general specifics" change to "without describing implementation details"

File: 007-design-docs.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+
+### Where to write a design document
+
+A design document should be a markdown file located in the `design` directory of the repository they are located in, for instance all design documents relating to the strimzi operator(s) should be located in the `strimzi-kafka-operator` repository under the `design` folder.
```
Comment (katheris) [inline 007-design-docs.md:43]: "they are located in" - should this be "the component code is located in"

File: 007-design-docs.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+#### Strimzi Design Document Organization
+Design documents should be splt into logical areas, for repositories that aren't `strimiz-kafka-operator` I suspect this would start as a single or just a few succinct documents.
```
Comment (katheris) [inline 007-design-docs.md:46]: "splt" -> "split"
"strimiz-kafka-operator" -> "strimzi-kafka-operator"

File: 007-design-docs.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+For design documentation that would cover the whole operator and not just any single custom resource I would suggest a file named `operator.md` this would describe behaviours not unique to any one specific reconcile loop or verticle and appropriate links to other files containing specific (but still high-level) differences between reconciles could be described there. 
+
+
+### How to write a design document
+
+The design documents would live in the related repository in the Strimzi Organization and would match the current code state. Naturally it is hard to allign the design documents and the code behaviour, but it would be the respinsibility of a reviewer to identify if any of the changes are impactful enough to potentially effect the design docs. Due to the below described simplicity of the design docs, if they are found to be out of date, it should be fairly simply to raise a pull request with the behavioural changes documented.
```
Comment (katheris) [inline 007-design-docs.md:53]: "allign" -> "align"
"respinsibility" -> "responsibility
"effect" -> "affect" or "have an effect on"

File: 007-design-docs.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+Is this intended to be a standalone component, are other components dependent on it, or is it dependent on other components.
+Is this intended for a single Kafka cluster or to interact with multiple.
+
+
+### When is a design document complete
+
+A design document is considered complete when the reviewers and maintainers believe the document to accurately reflect the implementation of the components behaviour. 
+Once approved the document is merged and is understood to be correct, any mistakes or changes to behaviour should be corrected in the form of a pull request alongside the code that changes the behavior.
+
+A lot of implementation issues, problems and compromises are only clear during or after development of a feature or change, so a design doc is recommended to be written alongside the code development, this helps in avoiding the writing of invalid design documents and help hilight potentially problematic elements of implementation or important holes in the testing strategy.
```
Comment (katheris) [inline 007-design-docs.md:80]: "hilight" -> "highlight"

I think given that the scope of these design docs are quite high level it could be argued that it would be valuable to write them up front when creating a new component. I could imagine the document being created as a PR initially and then only merged once the component is complete. It's up to you whether you take on this suggestion, but just thought I'd share my thoughts.

File: 007-design-docs.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+Once approved the document is merged and is understood to be correct, any mistakes or changes to behaviour should be corrected in the form of a pull request alongside the code that changes the behavior.
+
+A lot of implementation issues, problems and compromises are only clear during or after development of a feature or change, so a design doc is recommended to be written alongside the code development, this helps in avoiding the writing of invalid design documents and help hilight potentially problematic elements of implementation or important holes in the testing strategy. 
+
+Having the design document within the pull request allows contributors and reviewers to evolve the design while developing and during the review. This allows tweaks to the design, for current discussion and future reading, as well as to the code accordingly. This ensures the code and design docs evolve in parallel.
+
+
+## Summary of changes
+
+In an effort to improve where we are, the above guidelines are suggestions for future contributors but no enforced or mandated. This means that we will not place the entire burden on the first person to come along who wants to make a change in a given area.
```
Comment (katheris) [inline 007-design-docs.md:87]: "no enforced" -> "not enforced"

It might be clearer to simplify this, it seems like if it isn't enforced then design docs will never be written. Maybe just saying "In an effort to improve where we are, the above guidelines are suggested for future contributions that are either new components or significant changes to the design or behaviour of existing components. This could result in some components only being partially documented but over time design documents for existing components can be contributed."

Then I would be tempted to get rid of the rest of this section as you are sort of repeating yourself.

**Review by katheris — COMMENTED** [review]
I've noted the spelling mistakes and there are two areas where I would propose maybe rewording, but I think if you're happy then this is good to go.

File: 007-design-docs.md (hunk 1 above)
Comment (samuel-hawker) [reply 007-design-docs.md:5]: How so? What would you say is the main goal?

Comment (samuel-hawker) [reply 007-design-docs.md:5]: _The main goal of having design documents for Strimzi is so that behaviours of Strimzi as a whole and each of the components are well defined and are agreed upon._

Is what I am changing this to
